google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageRetryHandler.java

```diff
@@ -0,0 +1,176 @@
+import java.net.SocketTimeoutException;
+import javax.net.ssl.SSLException;
+
+/**
+ * Simple counter class to keep track of retry and reopen attempts when StorageExceptions are
+ * encountered. Handles sleeping between retry/reopen attempts, as well as throwing an exception
+ * when all retries/reopens are exhausted.
+ */
+public class CloudStorageRetryHandler {
+  private int retries;
+  private int reopens;
+  private long totalWaitTime; // in milliseconds
+  private final int maxRetries;
+  private final int maxReopens;
+
+  /**
+   * Create a CloudStorageRetryHandler with the maximum retries and reopens set to the same value.
+   *
+   * @param maxRetriesAndReopens value for both maxRetries and maxReopens
+   */
+  public CloudStorageRetryHandler(final int maxRetriesAndReopens) {
+    this.maxRetries = maxRetriesAndReopens;
+    this.maxReopens = maxRetriesAndReopens;
+  }
+
+  /**
+   * Create a CloudStorageRetryHandler with the maximum retries and reopens set to different values.
+   *
+   * @param maxRetries maximum number of retries
+   * @param maxReopens maximum number of reopens
+   */
+  public CloudStorageRetryHandler(final int maxRetries, final int maxReopens) {
+    this.maxRetries = maxRetries;
+    this.maxReopens = maxReopens;
+  }
+
+  /**
+   * @return number of retries we've performed
+   */
+  public int retries() {
+    return retries;
+  }
+
+  /**
+   * @return number of reopens we've performed
+   */
+  public int reopens() {
+    return reopens;
+  }
+
+  /**
+   * Checks whether we should retry, reopen, or give up.
+   *
+   * In the latter case it throws an exception (this includes the scenario where
+   * we exhausted the retry count).
+   *
+   * Otherwise, it sleeps for a bit and returns whether we should reopen.
+   * The sleep time is dependent on the retry number.
+   *
+   * @param exs caught StorageException
+   * @return True if you need to reopen and then try again. False if you can just try again.
+   * @throws StorageException if the exception is not retriable, or if you ran out of retries.
+   */
+  public boolean handleStorageException(final StorageException exs) throws StorageException {
+    // None of the retryable exceptions are reopenable, so it's OK to write the code this way.
+    if (isRetryable(exs)) {
+      handleRetryForStorageException(exs);
+      return false;
+    }
+    if (isReopenable(exs)) {
+      handleReopenForStorageException(exs);
+      return true;
+    }
+    throw exs;
+  }
+
+  /**
+   * If the exception is retriable, sleep for a bit and take note of the retry.
+   * Otherwise (or if out of attempts), throws a StorageException if we've exhausted all retries.
+   *
+   * The sleep time is dependent on the retry number.
+   *
+   * @param exs The StorageException error that prompted this retry attempt.
+   */
+  public void handleRetryForStorageException(final StorageException exs) throws StorageException {
+    if (!isRetryable(exs)) {
+      throw exs;
+    }
```

Comment:
Currently, `handleRetryForStorageException()` throws if the exception it's passed is not retryable, but `handleReopenForStorageException()` does not throw if the exception it's passed is not re-openable. This is a bit confusing/inconsistent.

My recommendation here would be to make both `handleRetryForStorageException()` and `handleReopenForStorageException()` private (forcing all clients to go through `handleStorageException()` instead), remove the `isRetryable()` check from `handleRetryForStorageException()`, and rely on `handleStorageException()` to check whether each error is retryable/reopenable and delegate to the right handler method.